### PR TITLE
feat(registry/ui): filters

### PR DESCRIPTION
**Pull request by harnyk**

### CODE REVIEW

File: registry/client/src/sharedLibs/List.js [hunk 1]
```diff
@@ -1,23 +1,13 @@
...
-    TextField,
-    EditButton,
-    FunctionField,
-} from 'react-admin'; // eslint-disable-line import/no-unresolved
-import {
-    Empty,
-    ListBulkActions,
-    ListActionsToolbar,
-} from '../components';
+import { Datagrid, List, SimpleList, TextField, EditButton, FunctionField } from 'react-admin'; // eslint-disable-line import/no-unresolved
```
Comment (blackrabbit99) [inline registry/client/src/sharedLibs/List.js:3]: please avoid import/no-unresolved looks scary.

File: registry/server/apps/routes/getApps.ts [hunk 2]
```diff
@@ -23,6 +39,24 @@ const getApps = async (req: Request, res: Response): Promise<void> => {
     if (filters.q) {
         query.where('name', 'like', `%${filters.q}%`);
     }
+    if (filters.domainId) {
+        if (filters.domainId === 'null') {
+            query.whereNotExists(function () {
```
Comment (blackrabbit99) [inline registry/server/apps/routes/getApps.ts:44]: Could you please move the code to the service because it is hard to read it.

**Review by blackrabbit99 — CHANGES_REQUESTED** [review]

File: registry/server/apps/routes/getApps.ts [hunk 3]
```diff
@@ -23,6 +39,24 @@ const getApps = async (req: Request, res: Response): Promise<void> => {
     if (filters.q) {
         query.where('name', 'like', `%${filters.q}%`);
     }
+    if (filters.domainId) {
```
Comment (stas-nc) [inline registry/server/apps/routes/getApps.ts:42]: Could you please add a test involving these queries (if it's not exits yet).

File: registry/server/sharedLibs/routes/getSharedLibs.ts [hunk 4]
```diff
@@ -5,11 +5,26 @@ import preProcessResponse from '../../common/services/preProcessResponse';
...
+const filtersSchema = Joi.object<Filters>({
+    name: Joi.string().optional(),
+});
+
+const getSharedLibs = async (req: RequestWithFilters<Filters>, res: Response): Promise<void> => {
     const query = db
         .selectVersionedRows(Tables.SharedLibs, 'name', EntityTypes.shared_libs, [`${Tables.SharedLibs}.*`])
         .from(Tables.SharedLibs);
+
+    if (req.filters?.name) {
```
Comment (stas-nc) [inline registry/server/sharedLibs/routes/getSharedLibs.ts:24]: Could you please add a test involving these queries (if it's not exits yet).

File: registry/server/templates/routes/getTemplates.ts [hunk 5]
```diff
@@ -1,15 +1,44 @@
...
+const filtersSchema = Joi.object<Filters>({
+    domainId: Joi.alternatives(Joi.number(), Joi.string().valid('null')).optional(),
+});
+
+const getTemplates = async (req: RequestWithFilters<Filters>, res: Response): Promise<void> => {
+    const query = db.selectVersionedRowsFrom<Template>(Tables.Templates, 'name', EntityTypes.templates, [
         `${Tables.Templates}.*`,
     ]);
+
+    if (req.filters?.domainId) {
```
[the remark this answers is not in the document]
Comment (harnyk) [reply registry/server/templates/routes/getTemplates.ts:24]: Sure thing. The PR is still draft, it's not finished

File: registry/server/apps/routes/getApps.ts (hunk 3 above)
Comment (harnyk) [reply registry/server/apps/routes/getApps.ts:42]: ✅

File: registry/server/templates/routes/getTemplates.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (harnyk) [reply registry/server/templates/routes/getTemplates.ts:36]: ⌛ done 1/2
Still not covered: "filter the templates by domain which has this template set in the template500 field"

[the remark this answers is not in the document]
Comment (harnyk) [reply registry/server/templates/routes/getTemplates.ts:36]: done ✅

File: registry/server/apps/routes/getApps.ts (hunk 3 above)
Comment (harnyk) [reply registry/server/apps/routes/getApps.ts:42]: done ✅

File: registry/client/src/sharedLibs/List.js (hunk 1 above)
Comment (harnyk) [reply registry/client/src/sharedLibs/List.js]: Sure, done.
BTW, i don't see any eslint-related checks in the repo 🤔

File: registry/server/sharedLibs/routes/getSharedLibs.ts (hunk 4 above)
Comment (harnyk) [reply registry/server/sharedLibs/routes/getSharedLibs.ts:24]: ✅

File: registry/server/templates/routes/getTemplates.ts [hunk 6]
```diff
@@ -1,15 +1,64 @@
...
+        .orderBy('name', 'asc');
+
+    if (req.filters?.id || req.filters?.name) {
+        const name = req.filters?.name ?? req.filters?.id;
+        ok(name);
+        const normalizedNames = Array.isArray(name) ? name : [name];
+        query.whereIn('name', normalizedNames);
+    }
+
+    if (req.filters?.domainId) {
```
Comment (wRLSS) [inline registry/server/templates/routes/getTemplates.ts:36]: matching for template500 seems to happen multiple times, utilize some constant for that?

File: registry/server/sharedLibs/routes/getSharedLibs.ts [hunk 7]
```diff
@@ -1,15 +1,30 @@
...
+        query.whereILike('name', `%${req.filters.name}%`);
```
[the remark this answers is not in the document]
Comment (harnyk) [reply registry/server/sharedLibs/routes/getSharedLibs.ts]: From MySQL docs:
> SQL pattern matching enables you to use _ to match any single character and % to match an arbitrary number of characters (including zero characters).

PostgreSQL and SQLite docs describe the same behavior.

Covered with tests too.

File: registry/client/src/sharedLibs/List.js [hunk 8]
```diff
@@ -26,19 +16,17 @@ const PostList = props => {
...
             perPage={25}
             actions={permissions === 'readonly' ? false : undefined}
+            filters={<ListFilter />}
             empty={<Empty />}
         >
             {isSmall ? (
                 <SimpleList
-                    primaryText={record => `@sharedLibrary/${record.name}`}
-                    secondaryText={record => record.spaBundle}
+                    primaryText={(record) => `@sharedLibrary/${record.name}`}
```
[the remark this answers is not in the document]
Comment (harnyk) [reply registry/client/src/sharedLibs/List.js:24]: As I am new to this project, I don't know the purpose of this prefixing, so I have no better idea than calling it `getPrefixedSharedLibName(name: string): string` or smth like this.

File: registry/tests/templates.spec.ts [hunk 9]
```diff
@@ -273,7 +273,7 @@ describe(`Tests ${example.url}`, () => {
             }
         });
 
-        it('should filter a list of templates by domain', async () => {
+        it.only('should filter a list of templates by domainId using', async () => {
```
Comment (blackrabbit99) [inline registry/tests/templates.spec.ts]: please avoid only and if possible catch it in lint

**Review by blackrabbit99 — CHANGES_REQUESTED** [review]

File: registry/tests/templates.spec.ts (hunk 9 above)
Comment (harnyk) [reply registry/tests/templates.spec.ts]: sure thing. BTW, it's already removed in the recent commits.

I will create an issue for adding eslint (we don't have it in `ilc` AFAIK).

File: registry/server/templates/routes/getTemplates.ts (hunk 6 above)
Comment (harnyk) [reply registry/server/templates/routes/getTemplates.ts:36]: It's a field name, just like "name" or "domainId". If we put it into constant, we will need to do the same for the rest of the fields, just like we do for tables.

File: registry/server/templates/routes/getTemplates.ts [hunk 10]
```diff
@@ -1,15 +1,56 @@
...
+
+const getTemplates = async (req: RequestWithFilters<Filters>, res: Response): Promise<void> => {
+    const query = db.selectVersionedRowsFrom<Template>(Tables.Templates, 'name', EntityTypes.templates, [
         `${Tables.Templates}.*`,
     ]);
+
+    if (req.filters?.id || req.filters?.name) {
+        const name = req.filters?.name ?? req.filters?.id;
+        ok(name);
+        const normalizedNames = Array.isArray(name) ? name : [name];
```
[the remark this answers is not in the document]
Comment (harnyk) [reply registry/server/templates/routes/getTemplates.ts:32]: In TS it becomes:

```ts
const normalizedName = ([] as string[]).concat(name)
```

But I still prefer the more literal way for its readability.

File: registry/server/templates/routes/getTemplates.ts [hunk 11]
```diff
@@ -1,21 +1,19 @@
...
+    domainId: Joi.alternatives(Joi.number(), Joi.string().valid('null')).optional(),
+    id: Joi.alternatives(Joi.string(), Joi.array().items(Joi.string())).optional(),
+    name: Joi.alternatives(Joi.string(), Joi.array().items(Joi.string())).optional(),
+});
 
-    res.setHeader('Content-Range', templates.length); //Stub for future pagination capabilities
-    res.status(200).send(itemsWithId);
+const getTemplates = async (req: RequestWithFilters<TemplatesGetListFilters>, res: Response): Promise<void> => {
+    const { data, pagination } = await templatesRepository.getList(req.filters ?? {});
+    res.setHeader('Content-Range', pagination.total); //Stub for future pagination capabilities
```
Comment (yehor-manzhula) [inline registry/server/templates/routes/getTemplates.ts:15]: I would not add that until it is required to be present

File: registry/server/sharedLibs/routes/getSharedLibs.ts [hunk 12]
```diff
@@ -1,22 +1,21 @@
...
+});
+
+const getSharedLibs = async (req: RequestWithFilters<SharedLibsGetListFilters>, res: Response): Promise<void> => {
+    const { data: sharedLibs, pagination } = await sharedLibsRepository.getList(req.filters ?? {}, {
+        range: req.query.range as string | undefined,
     });
 
-    res.setHeader('Content-Range', sharedLibs.pagination.total);
-    res.status(200).send(preProcessResponse(itemsWithId));
+    res.setHeader('Content-Range', pagination.total);
```
Comment (yehor-manzhula) [inline registry/server/sharedLibs/routes/getSharedLibs.ts:17]: Same here

File: registry/server/templates/routes/getTemplates.ts (hunk 11 above)
Comment (harnyk) [reply registry/server/templates/routes/getTemplates.ts:15]: I left it from the implementation made by previous developer. The client-side code relies on it

**Review by wRLSS — APPROVED** [review]

**Pull request merged**